openedx/core/djangoapps/schedules/resolvers.py

```diff
@@ -1,136 +1,313 @@
+                'student_name': user.profile.name,
+
+                'course_name': first_schedule.enrollment.course.display_name,
+                'course_url': absolute_url(self.site, reverse('course_root', args=[str(first_schedule.enrollment.course_id)])),
+
+                # This is used by the bulk email optout policy
+                'course_ids': course_id_strs,
+            })
+
+            # Information for including upsell messaging in template.
+            _add_upsell_button_information_to_template_context(
+                user, first_schedule, template_context)
+
+            yield (user, first_schedule.enrollment.course.language, template_context)
+
+
+def _get_datetime_beginning_of_day(dt):
+    """
+    Truncates hours, minutes, seconds, and microseconds to zero on given datetime.
+    """
+    return dt.replace(hour=0, minute=0, second=0, microsecond=0)
 
 
 class UpgradeReminderResolver(BinnedSchedulesBaseResolver):
     """
     Send a message to all users whose verified upgrade deadline is at ``self.current_date`` + ``day_offset``.
     """
-    async_send_task = upgrade_reminder_schedule_bin
+    log_prefix = 'Upgrade Reminder'
+    schedule_date_field = 'upgrade_deadline'
     num_bins = UPGRADE_REMINDER_NUM_BINS
-    enqueue_config_var = 'enqueue_upgrade_reminder'
 
-    def __init__(self, *args, **kwargs):
-        super(UpgradeReminderResolver, self).__init__(*args, **kwargs)
-        self.log_prefix = 'Upgrade Reminder'
+    def schedules_for_bin(self):
+        schedules = self.get_schedules_with_target_date_by_bin_and_orgs()
+
+        for (user, user_schedules) in groupby(schedules, lambda s: s.enrollment.user):
+            user_schedules = list(user_schedules)
+            course_id_strs = [str(schedule.enrollment.course_id) for schedule in user_schedules]
+
+            first_schedule = user_schedules[0]
+            template_context = get_base_template_context(self.site)
+            template_context.update({
+                'student_name': user.profile.name,
+                'user_personal_address': user.profile.name if user.profile.name else user.username,
+
+                'course_links': [
+                    {
+                        'url': absolute_url(self.site, reverse('course_root', args=[str(s.enrollment.course_id)])),
+                        'name': s.enrollment.course.display_name
+                    } for s in user_schedules
+                ],
+
+                'first_course_name': first_schedule.enrollment.course.display_name,
+
+                # This is used by the bulk email optout policy
+                'course_ids': course_id_strs,
+
+                'cert_image': absolute_url(self.site, static('course_experience/images/verified-cert.png')),
+            })
+
+            _add_upsell_button_information_to_template_context(user, first_schedule, template_context)
+
+            yield (user, first_schedule.enrollment.course.language, template_context)
```

Comment:
Looks like this could be even further DRYed with `ScheduleStartResolver` since it shares exactly the same logic.  Only the class variables and a few template fields are different.  Even the template fields could be shared and consolidated.

Reply:
Yeah, I considered that, and then wasn't sure whether that would have been taking it too far. I'll look again and see what a reasonable balance is.